probes/memorysafe/impl.go

```diff
@@ -0,0 +1,229 @@
+func init() {
+	probes.MustRegisterIndependent(Probe, Run)
+}
+
+func Run(raw *checker.CheckRequest) (found []finding.Finding, probeName string, err error) {
+	prominentLangs := getRepositoryLanguageChecks(raw)
+	findings := []finding.Finding{}
+
+	for _, lang := range prominentLangs {
+		for _, langFunc := range lang.funcPointers {
+			if langFunc == nil {
+				raw.Dlogger.Warn(&checker.LogMessage{
+					Text: fmt.Sprintf("no function pointer found for language %s", lang.Desc),
+				})
+			}
+			langFindings, err := langFunc(raw)
+			if err != nil {
+				return nil, Probe, fmt.Errorf("error while running function for language %s: %w", lang.Desc, err)
+			}
+			findings = append(findings, langFindings...)
+		}
+	}
+	return findings, Probe, nil
+}
+
+func getRepositoryLanguageChecks(raw *checker.CheckRequest) []languageMemoryCheckConfig {
+	langs, err := raw.RepoClient.ListProgrammingLanguages()
+	if err != nil {
+		raw.Dlogger.Warn(&checker.LogMessage{
+			Text: fmt.Sprintf("RepoClient retured error for ListProgrammingLanguages: %v", err),
+		})
+		return nil
+	}
+	if len(langs) == 0 {
+		return []languageMemoryCheckConfig{}
+	}
+	if len(langs) == 1 && langs[0].Name == clients.All {
+		return getAllLanguages()
+	}
+	ret := []languageMemoryCheckConfig{}
+	for _, language := range langs {
+		if lang, ok := languageMemorySafeSpecs[clients.LanguageName(strings.ToLower(string(language.Name)))]; ok {
+			ret = append(ret, lang)
+		}
+	}
+	return ret
+}
+
+func getAllLanguages() []languageMemoryCheckConfig {
+	allLanguages := make([]languageMemoryCheckConfig, 0, len(languageMemorySafeSpecs))
+	for l := range languageMemorySafeSpecs {
+		allLanguages = append(allLanguages, languageMemorySafeSpecs[l])
+	}
+	return allLanguages
+}
+
+// Golang
+
+func checkGoUnsafePackage(client *checker.CheckRequest) ([]finding.Finding, error) {
+	findings := []finding.Finding{}
+
+	if err := fileparser.OnMatchingFileContentDo(client.RepoClient, fileparser.PathMatcher{
+		Pattern:       "*.go",
+		CaseSensitive: false,
+	}, goCodeUsesUnsafePackage, &findings, client.Dlogger); err != nil {
+		return nil, err
+	}
+	if len(findings) == 0 {
+		found, err := finding.NewWith(fs, Probe,
+			"Golang code does not use the unsafe package", nil, finding.OutcomeTrue)
+		if err != nil {
+			return nil, fmt.Errorf("create finding: %w", err)
+		}
+		findings = append(findings, *found)
+	}
+	return findings, nil
+}
+
+func goCodeUsesUnsafePackage(path string, content []byte, args ...interface{}) (bool, error) {
+	findings, ok := args[0].(*[]finding.Finding)
+	if !ok {
+		// panic if it is not correct type
+		panic(fmt.Sprintf("expected type findings, got %v", reflect.TypeOf(args[0])))
+	}
+	fset := token.NewFileSet()
+	f, err := parser.ParseFile(fset, "", content, parser.ImportsOnly)
+	if err != nil {
+		dl, ok := args[1].(checker.DetailLogger)
+		if !ok {
+			// panic if it is not correct type
+			panic(fmt.Sprintf("expected type checker.DetailLogger, got %v", reflect.TypeOf(args[1])))
```

Comment:
we should move the type checking out of the `err != nil` block to make the panics when passing wrong types more obvious.